Implement EditAction for keyboard gestures

pyroxymat opened the pull request:
This moves our textbox actions into an `EditAction` enum with the goal of eventually moving to a xi-based implementation for our textbox.

This is a step in solving issue #615

futurepaul commented:
This already feels like a big improvement to me! I don't know exactly where this is headed, so the `SingleLineTextInput` / `TextInput` abstraction feels a little premature to me, but I'll defer to @cmyr on this.

And a potential point of clarification: the way I've been thinking of text's future in druid is making it so that xi's `Rope` data structure would be easy to swap in to replace `String` if necessary. The other thinking is that the text editing facilities could be useful to other UI toolkits, so should stay generic if possible. I've ignored the broader abstractions of xi editor, because afaik it's not obvious that a simple text editor in druid should use RPC, for instance.

futurepaul commented:
It was annoying but we were resetting the cursor blink everywhere it made sense. I believe this is a regression in that regard. Is there a slick way to get that back without being messy?

futurepaul commented:
Movements don't seem to be updating the hscroll. When I jump left or jump right on a line that's longer than the textbox, the cursor goes but the hscroll doesn't. Same when I use the arrow keys to move through a piece of text longer than textbox.

pyroxymat commented:
> It was annoying but we were resetting the cursor blink everywhere it made sense. I believe this is a regression in that regard. Is there a slick way to get that back without being messy?

I updated the logic that no matter which "EditAction" we currently execute, we always reset the Blinking. Notice that an edit action will always cause a change in either selection or text. If this changes we would need to re-evaluate

> Movements don't seem to be updating the hscroll. When I jump left or jump right on a line that's longer than the textbox, the cursor goes but the hscroll doesn't. Same when I use the arrow keys to move through a piece of text longer than textbox.

I changed it that any movement (except for select all) causes the hscroll to be updated. This adds the really nice feature (that wasn't in before) that if drag select more than currently fits in the textbox it will automatically scroll to track your cursor.

cmyr commented on druid/src/text/text_input.rs:
```diff
@@ -0,0 +1,96 @@
+// This is done with the goal of eventually being able to easily switch
+// to a xi-based implementation of our Events.
+
+/// Events that only modify view state
```
This comment is now out of date

pyroxymat replied:
Changed!

cmyr commented on druid/src/text/text_input.rs:
```diff
@@ -0,0 +1,96 @@
+use super::Movement;
```
can we get a license header and a `//! short module comment` here?

pyroxymat replied:
Done

cmyr commented on druid/src/text/text_input.rs:
```diff
@@ -0,0 +1,96 @@
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct MouseAction {
```
I'd like doc comments on all public items please. :)

Also: who is doing click detection?

pyroxymat replied:
The textbox itself

cmyr commented on druid/src/text/text_input.rs:
```diff
@@ -0,0 +1,113 @@
+// This is done with the goal of eventually being able to easily switch
+// to a xi-based implementation of our Events.
+
+/// Defines actions that edit the state of a textbox widget
```
```suggestion
/// An enum that represents actions in a text buffer.
```

cmyr commented on druid/src/text/text_input.rs:
```diff
@@ -0,0 +1,113 @@
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct MouseAction {
+    pub column: usize,
+    pub shift: bool,
```
I think this should be row, column, and `KeyModifiers`. We do care about ctrl/cmd/etc.

cmyr commented on druid/src/text/text_input.rs:
```diff
@@ -0,0 +1,113 @@
+
+/// Handles key events and returns actions that are applicable to
+/// single line textboxes
+pub struct SingleLineTextInput {}
```
We don't need to overthink this; single line input is a subset of normal input, and doesn't really require special logic; we just don't handle certain events later on. I would just call this `BasicTextInput` or something. (We will want some platform-specific additions, as well, like `CocoaTextInput`; we can hopefully chain these, like): 

```rust
let edit = basic_input.handle_event(&event).or(platform_input.handle_event(&event));
```

... or something.

pyroxymat replied:
Seems like a smart idea. Fixed.

cmyr commented:
@pyroxymat ping on this? very interested in merging.
